geps/gep-1619.md

```diff
@@ -74,28 +74,122 @@ While session persistence uses attributes in the application layer, session affi
 ### Sessions in Java
 
 Java application servers such as Tomcat and Jetty, were the first to standardize the API around persistent sessions. These Java applications introduced the “jsessionid” cookie as well as more advanced features such as session migration, replication, and on demand session activation. It’s important for Gateway API to examine persistent session use cases and history from Java APIs to ensure the API is designed appropriately.
 
+## API
+
+### Session Persistence API
+
+The majority of implementations and dataplanes support configuring cookies for session persistence while support for header-based session persistence is limited.
+
+To organize potential Gateway API configuration of cookie functionality, the next section will break down configuration into specific categories:
+- Session Cookie Configuration
+- Session Cookie Strategy
+
+**Session Cookie Configuration**
+
+Session cookie configuration are the parameters and attributes the dataplane uses to create a cookie for establishing a session.
+
+A set-cookie request can be broken down into [3 main components](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie):
+- Cookie Name
+- Cookie Value
+- Cookie Attributes
+
+Only the Cookie Name and Cookie Value are required. All Cookie Attributes are optional. Attributes are simply key=value pairs with the value as optional for some attributes.
+
+The set-cookies attributes defined by [rfc6265](https://www.rfc-editor.org/rfc/rfc6265#section-5.2) are:
+- Expires=_date_
+- Max-Age=_number_
+- Domain=_domain_
+- Path=_path-value_
+- Secure
+- HttpOnly
+
+However, there are de-facto standard attributes that have been proposed as updates to [rfc6265](https://www.rfc-editor.org/rfc/rfc6265#section-5.2):
+- SameSite=_[Strict|Lax|None]_ ([Draft RFC Update](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-cookie-same-site-00))
+- Partitioned ([Draft RFC Update](https://www.ietf.org/archive/id/draft-cutler-httpbis-partitioned-cookies-01.html))
+
+**Session Cookie Strategy**
+
+Session cookie strategy is configuration specific to the creation of the cookie by the dataplane. These are not specific configuration parameters of the cookie itself, but instead the operation to take when inserting the cookie such as:
+- Rewriting an existing cookie value
+- Inserting a new cookie
+- Prefixing an existing cookie
+- Preserving an existing cookie value
+- Only adding cookies on specific HTTP methods (POST/GET)
+
+### Session Affinity API
+
+TBD - See open question about session affinity
+
 ### Open Questions
 
-TBD
+- Should we include both session persistence and session affinity API configurations in this GEP or just focus on session persistence?
```

Comment:
Another option is to not differentiate between persistence and affinity and combine both into a single concept.

Reply:
At a minimum, I agree with redefining "session affinity". It's not immediately clear to me, based on the words themselves, what the difference between persistence and affinity is, and lots of implementations use them interchangeably and I think this would be confusing if not changed. But I've heard differing thoughts on whether we should still keep the idea of strong session persistence (cookies/headers) and weak session persistence (connection attribute hashing). 

John mentioned this in the first PR review: https://github.com/kubernetes-sigs/gateway-api/pull/1643#discussion_r1068250615. Per the comment, I developed the definition with his comment, but kept the labels of persistence/affinity:

> In my view, the primary difference between session types is whether the client directly specifies the backend they want OR if some attribute of the request is just used to consistently send to a backend.

Here are my thoughts I'm working through right now:
- If our future API for persistence and affinity is designed in such as way that the configurations live together, such as Istio's `ConsistentHashLB`, the to me, it makes sense we can redefine the concept of persistence and affinity into the same "session persistence" bucket.
  - However, grouping loadbalancing algorithms in with session persistence feel awkward to me. Some loadbalancing algorithms provide consistent hashing, others don't. Thinking ahead, GWAPI could potentially support configuring the algorithm in the future, but if some algorithms are selected in the session persistence API object, then how would you configure the non-consistent hashing algorithms? I think we'd have to separate to into something like [SimpleLB](https://istio.io/latest/docs/reference/config/networking/destination-rule/#LoadBalancerSettings-SimpleLB) and  [ConsistentHashLB](https://istio.io/latest/docs/reference/config/networking/destination-rule/#LoadBalancerSettings-ConsistentHashLB) like Istio. I think these concepts of "LBs" might be hard to translate to GWAPI. Though, we'd still have this problem if we had a API specifically for session affinity.
  - The other downside, is that we lose the distinction of guaranteed session persistence (headers/cookies) vs. best attempt (using connection properties, or hashing that may change or rebalance upon new backend)
- If our future API will keep persistence and affinity as separate objects, then it makes sense to define them differently (current state of this GEP).
  - May be more confusing vs. the value added for the separation.

So I'd like to start a discussion, involving @costinm @howardjohn and whoever has input, on how we should define these concepts. I lean towards the simplicity of calling everything session persistence.

Reply:
Thinking more about it I agree that there are 2 concepts and I would call them session persistence and affinity (without "session").

- session persistence: there is a concept called session where a client creates a session by sending the first request in a session without any special attribute (e.g. cookie). The infrastructure (proxy/LB/gateway etc) routes this "first request" based on applicable policies. The response contains the attribute (cookie) which the client includes in all the subsequent requests in the session thereby explicitly indicating the session semantics. The infrastructure respects the cookie to route the requests to the same backend as the first one thereby maintaining session persistence.

- affinity: this just involves the infrastructure (LB/proxy etc) using the request attributes (e.g. a header) to select a backend thereby maintaining affinity based on those attributes. Although the client application can simulate a "session" by injecting that header with a specific value, there is no special first request initiating a session.

So I am okay with keeping the existing terminology (session persistence and session affinity)

Reply:
That's valid. Just to be clear, are you saying to drop the "session" from "session affinity" for just "affinity"? Or keep it as "session affinity"?

I now realize Envoy defines them as [stateful sessions](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/http/stateful_session/cookie/v3/cookie.proto) as so does your [gRPC proposal](https://github.com/grpc/proposal/blob/master/A55-xds-stateful-session-affinity.md#background). The word "stateful" is used to distinguish between cookie-based session persistence and consistent hashing. That's another idea.

Reply:
Okay - got it. Costin mentioned https://kubernetes.io/docs/reference/networking/virtual-ips/#session-affinity. Looks like Kubernetes refers to clientIP-based consistent sessions as "session affinity", which technically aligns to our naming conventions, but Kubernetes doesn't go as far to define that persistence != affinity as we are trying to.

I think the designations of persistence and affinity make sense, but I will most likely bring up in one our meetings to get feedback. Sounds like we agree on defining two separate things, but it's probably worth making sure there isn't a better naming convention (say stateful sessions vs stateless sessions or strong session persistence vs. weak session persistence). I worry "session affinity" is overloaded.

If we end up designing the persistence and affinity APIs to be tightly coupled, say LoadbalancerPolicy-centric design (like you are advocating for in https://github.com/kubernetes-sigs/gateway-api/pull/1935#issuecomment-1525438073), we might never need to distinguish between the two (they all be in the same bucket).

Reply:
@gcs278 can you capture some of this context in the GEP before merging and link to this thread?

Reply:
Done: https://github.com/kubernetes-sigs/gateway-api/pull/1935/files#diff-afa0e827962fe48152635217c6bc954020d5fec3578af4a9e5f744a2b4568fffR80-R89

There were a couple ideas floating in this thread, but I added a section on the relationship of session persistence and session affinity as we have discussed it in the PR.